src/opnsense/mvc/app/controllers/OPNsense/Base/ApiControllerBase.php

```diff
@@ -68,6 +68,63 @@ private function parseJsonBodyData()
         return null;
     }
 
+    private function recursiveFilter($records, $searchPhrase)
+    {
+        foreach($records as $val) {
+            if (is_array($val)) {
+                return $this->recursiveFilter($val, $searchPhrase);
+            }
+            elseif (stripos($val, $searchPhrase) !== false) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Generic legacy search action, reads post variables for filters and page navigation.
+     * @param $records An array of records to process for filter and pagination
+     * @return array
+     */
+    protected function applyPaginationFilter($records)
+    {
+        $itemsPerPage = intval($this->request->getPost('rowCount', 'int', 9999));
+        $currentPage = intval($this->request->getPost('current', 'int', 1));
+        $offset = ($currentPage - 1) * $itemsPerPage;
+
+        // Sort entries
+        if ($this->request->hasPost('sort') && is_array($this->request->getPost('sort'))) {
+            $keys = array_keys($this->request->getPost('sort'));
+            $order = $this->request->getPost('sort')[$keys[0]];
+            $keys = array_column($records, $keys[0]);
+            if (count($keys) === count($records)) {
+                array_multisort($keys, $order == 'asc' ? SORT_ASC : SORT_DESC, $records);
+            }
+        }
+
+        // Filter entries
+        $entry_keys = array_keys($records);
+        if ($this->request->hasPost('searchPhrase') && $this->request->getPost('searchPhrase') !== '') {
+            $searchPhrase = (string)$this->request->getPost('searchPhrase');
+            $entry_keys = array_filter($entry_keys, function ($key) use ($searchPhrase, $records) {
+                return $this->recursiveFilter($records[$key], $searchPhrase);
```

Comment:
I'm a bit doubting we should try to parse these records recursively to be honest, my initial thought was to bring the existing code (from TunnelController.php) into here (optionally with a parameter to specify which fields to search). The downside from filtering structures is that results will be returned which are unlikely to be returned in the view.

Reply:
moved the generic search code in https://github.com/opnsense/core/commit/2e747acda90d447f4047c5e7bce941fb2f8bf5cb